File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/SortList.java
package com.only4.algorithm.version4j.leetcode;

import com.only4.algorithm.version4j.extra.ListNode;

public class SortList {
    public ListNode sortList(ListNode head) {
        if (head == null || head.next == null) return head;

        ListNode middleNode = findMiddleAndSplit(head);
        ListNode leftSorted = sortList(head);
        ListNode rightSorted = sortList(middleNode);

        return mergeTwoSortedLists(leftSorted, rightSorted);
    }

    private ListNode findMiddleAndSplit(ListNode head) {
        ListNode slowPointer = head;
        ListNode fastPointer = head.next;

        while (fastPointer != null && fastPointer.next != null) {
            slowPointer = slowPointer.next;
            fastPointer = fastPointer.next.next;
        }

        ListNode middleNode = slowPointer.next;
        slowPointer.next = null; // 分割链表
        return middleNode;
    }

    private ListNode mergeTwoSortedLists(ListNode firstList, ListNode secondList) {
        ListNode dummyHead = new ListNode(0);
        ListNode currentNode = dummyHead;

        while (firstList != null && secondList != null) {
            if (firstList.val <= secondList.val) {
                currentNode.next = firstList;
                firstList = firstList.next;
            } else {
                currentNode.next = secondList;
                secondList = secondList.next;
            }
            currentNode = currentNode.next;
        }

        currentNode.next = (firstList != null) ? firstList : secondList;
        return dummyHead.next;
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/sortList.kt
package com.only4.algorithm.version4k.leetcode

import com.only4.algorithm.version4k.extra.ListNode

fun sortList(head: ListNode?): ListNode? {
    if (head?.next == null) return head

    fun findMiddleAndSplit(head: ListNode): ListNode {
        var slowPointer = head
        var fastPointer = head.next

        while (fastPointer?.next != null) {
            slowPointer = slowPointer.next!!
            fastPointer = fastPointer.next?.next
        }

        val middleNode = slowPointer.next!!
        slowPointer.next = null // 分割链表
        return middleNode
    }

    fun mergeTwoSortedLists(firstList: ListNode?, secondList: ListNode?): ListNode? {
        val dummyHead = ListNode(0)
        var currentNode = dummyHead
        var leftPointer = firstList
        var rightPointer = secondList

        while (leftPointer != null && rightPointer != null) {
            if (leftPointer.`val` <= rightPointer.`val`) {
                currentNode.next = leftPointer
                leftPointer = leftPointer.next
            } else {
                currentNode.next = rightPointer
                rightPointer = rightPointer.next
            }
            currentNode = currentNode.next!!
        }

        currentNode.next = leftPointer ?: rightPointer
        return dummyHead.next
    }

    val middleNode = findMiddleAndSplit(head)
    val leftSorted = sortList(head)
    val rightSorted = sortList(middleNode)

    return mergeTwoSortedLists(leftSorted, rightSorted)
}
